mapic/mistapiconnector_app.go

```diff
@@ -423,38 +445,113 @@ func (mc *mac) shouldEnableAudio(stream *api.Stream) bool {
 	return audio
 }
 
-func (mc *mac) startMultistream(wildcardPlaybackID, playbackID string, info *streamInfo) {
-	for i := range info.stream.Multistream.Targets {
-		go func(targetRef api.MultistreamTargetRef) {
-			glog.Infof("==> starting multistream %s", targetRef.ID)
-			target, pushURL, err := mc.getPushUrl(info.stream, &targetRef)
-			if err != nil {
-				glog.Errorf("Error building multistream target push URL. targetId=%s stream=%s err=%v",
-					targetRef.ID, wildcardPlaybackID, err)
-				return
-			} else if target.Disabled {
-				glog.Infof("Ignoring disabled multistream target. targetId=%s stream=%s",
-					targetRef.ID, wildcardPlaybackID)
-				return
+// reconcileMultistreamLoop calls reconcileMultistream periodically or when multistreamUpdated is triggered on demand.
+func (mc *mac) reconcileMultistreamLoop(ctx context.Context) {
+	ticker := time.NewTicker(1 * time.Minute)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		case <-mc.multistreamUpdated:
+		}
+		mc.reconcileMultistream()
+	}
+}
+
+// reconcileMultistream makes sure that Mist contains the multistream pushes exactly as specified in streamInfo cache.
+// There may be multiple reasons why Mist is not in sync with streamInfo cache:
+// - streamInfo cache has changed (multistream target was turned on/off or multistream target was added/removed)
+// - Mist removed its push for some reason
+// Note that we use Mist AUTO_PUSH (which in turn makes sure that the PUSH is always available).
+func (mc *mac) reconcileMultistream() {
+	type key struct {
+		stream string
+		target string
+	}
+	toKey := func(stream, target string) key {
+		return key{stream: stream, target: target}
+	}
+	isMultistream := func(k key) bool {
+		return (strings.HasPrefix(k.stream, "video+") || strings.HasPrefix(k.stream, "videorec+")) &&
+			(strings.HasPrefix(strings.ToLower(k.target), "rtmp:") || strings.HasPrefix(strings.ToLower(k.target), "srt:"))
+	}
+
+	// Get the existing PUSH_AUTO from Mist
+	mistPushAutoList, err := mc.mist.PushAutoList()
+	if err != nil {
+		glog.Errorf("error executing PUSH_AUTO_LIST on Mist, cannot reconcile multistream err=%v", err)
+		return
+	}
+	var filteredMistPushAutoList []clients.MistPushAuto
+	mistMap := map[key]bool{}
+	for _, e := range mistPushAutoList {
+		k := toKey(e.Stream, e.Target)
+		if isMultistream(k) {
```

Comment:
What other things does Mist return here which are not multistream? Is it the recordings bucket?

Reply:
Currently, apart from multistream, auto push is used for recordings.